pkg/controller/clusterpool/clusterpool_controller.go

```diff
@@ -415,6 +424,128 @@ func (r *ReconcileClusterPool) Reconcile(ctx context.Context, request reconcile.
 	return reconcile.Result{}, nil
 }
 
+// updateInventory ensures that the inventory of the cluster pool and the related ClusterDeploymentCustomizations are up to date.
+// This includes validating resources existance and updating references when needed.
+func (r *ReconcileClusterPool) updateInventory(clp *hivev1.ClusterPool, cds []*hivev1.ClusterDeployment, status hivev1.LastApplyStatusType, logger log.FieldLogger) {
+	// no need to update Inventory if it doesn't exist
+	if clp.Spec.Inventory == nil {
+		return
+	}
+	// InventoryValid condition and reason
+	reason := hivev1.InventoryReasonValid
+	switch status {
+	case hivev1.LastApplyBrokenCloud:
+		reason = hivev1.InventoryReasonBrokenByCloud
+	case hivev1.LastApplyBrokenSyntax:
+		reason = hivev1.InvenotryReasonBrokenBySyntax
+	}
+
+	// Helper functions to get and update ClusterDeploymentCustomization, and update inventory message if missing
+	getCDC := func(name string) *hivev1.ClusterDeploymentCustomization {
+		cdc := &hivev1.ClusterDeploymentCustomization{}
+		if err := r.Client.Get(context.Background(), client.ObjectKey{Namespace: clp.Namespace, Name: name}, cdc); err != nil {
+			if apierrors.IsNotFound(err) {
+				reason := hivev1.InventoryReasonMissing
+				r.updateInventoryValidMessage(clp, cdc.Name, reason, false, logger)
+			}
+			log.WithError(err).Warn("error reading customization")
+			return nil
+		}
+		return cdc
+	}
+
+	updateCDC := func(cdc *hivev1.ClusterDeploymentCustomization, status bool, msg string) {
+		if err := r.setCustomizationAvailabilityCondition(cdc, status, msg, logger); err != nil {
+			log.WithError(err).Warn("failed to update customization status")
+		}
+	}
+
+	// First update the inventory and ClusterDeploymentCustomizations related to changed ClusterDeployments
+	var active_cdc []string
+	for _, cd := range cds {
+		active_cdc = append(active_cdc, cd.Spec.ClusterPoolRef.CustomizationRef.Name)
+		if cdc := getCDC(cd.Spec.ClusterPoolRef.CustomizationRef.Name); cdc != nil {
+			r.updateInventoryValidMessage(clp, cdc.Name, reason, false, logger)
+			// Fix missing ClusterDeployment Reference
+			if cdc.Status.ClusterDeploymentRef == nil {
+				cdc.Status.ClusterDeploymentRef = &corev1.LocalObjectReference{Name: cd.Name}
+			}
+
+			cdc.Status.LastApplyStatus = status
+
+			updateCDC(cdc, false, "reserved")
+		}
+	}
+	r.updateInventoryValidMessage(clp, "", "", true, logger) // just the inventory condition is being updated
+	sort.Strings(active_cdc)
+
+	// Next update the rest of the inventory and ClusterDeploymentCustomizations
+	for _, item := range clp.Spec.Inventory {
+		pos := sort.SearchStrings(active_cdc, item.Name)
+		if !(pos < len(active_cdc) && active_cdc[pos] == item.Name) {
+			continue
+		}
```

Comment:
This is still a weird way to [search for an element in a slice](https://freshman.tech/snippets/go/check-if-slice-contains-element/).